### PR TITLE
tstest/natlab/vnet: add start of virtual network-based NAT Lab

### PR DESCRIPTION
**Pull request by bradfitz**
Updates #13038

### CODE REVIEW

File: cmd/vnet/run-krazy.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+
+echo "Type 'C-a c' to enter monitor; q to quit."
+
+set -eux
+qemu-system-x86_64 -M microvm,isa-serial=off \
+    -m 1G \
+    -nodefaults -no-user-config -nographic \
+    -kernel $HOME/src/github.com/tailscale/gokrazy-kernel/vmlinuz \
+    -append "console=hvc0 root=PARTUUID=60c24cc1-f3f9-427a-8199-dd02023b0001/PARTNROFF=1 ro init=/gokrazy/init panic=10 oops=panic pci=off nousb tsc=unstable clocksource=hpet" \
```
Comment (maisem) [inline cmd/vnet/run-krazy.sh:10]: random GUID?

File: cmd/vnet/run-krazy.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    -drive id=blk0,file=$HOME/src/tailscale.com/gokrazy/tsapp.img,format=raw \
+    -device virtio-blk-device,drive=blk0 \
+    -netdev stream,id=net0,addr.type=unix,addr.path=/tmp/qemu.sock \
+    -device virtio-serial-device \
+    -device virtio-net-device,netdev=net0,mac=52:cc:cc:cc:cc:00 \
```
Comment (maisem) [inline cmd/vnet/run-krazy.sh:15]: don't we need to change this based on the inputs?

File: tstest/natlab/vnet/conf.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+//
+// On an error or unknown opt type, AddNode returns a
+// node with a carried error that gets returned later.
+func (c *Config) AddNode(opts ...any) *Node {
+	num := len(c.nodes)
+	n := &Node{
+		mac: MAC{0x52, 0xcc, 0xcc, 0xcc, 0xcc, byte(num)}, // 52=TS then 0xcc for ccclient
+	}
+	c.nodes = append(c.nodes, n)
+	for _, o := range opts {
```
Comment (maisem) [inline tstest/natlab/vnet/conf.go:44]: maybe restrict to one network for now as thats the only thing that works?

File: cmd/vnet/run-krazy.sh (hunk 1 above)
Comment (bradfitz) [reply cmd/vnet/run-krazy.sh:10]: that's what my gokrazy build generated. Don't know if it's fixed or not.

This script is just an example/demo. It'll be deleted later.

File: cmd/vnet/run-krazy.sh (hunk 2 above)
Comment (bradfitz) [reply cmd/vnet/run-krazy.sh:15]: This script is just an example used to show how fast gokrazy booting is with microvm and to quickly hand check that natlab works. This script will be deleted.

File: tstest/natlab/vnet/conf.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+			return fmt.Errorf("two nodes have the same MAC %v", n.mac)
+		}
+		s.nodes = append(s.nodes, n)
+		s.nodeByMAC[n.mac] = n
+
+		// Allocate a lanIP for the node. Use the network's CIDR and use final
+		// octet 101 (for first node), 102, etc. The node number comes from the
+		// last octent of the MAC address (0-based)
+		ip4 := n.net.lanIP.Addr().As4()
+		ip4[3] = 101 + n.mac[5]
```
Comment (maisem) [inline tstest/natlab/vnet/conf.go:200]: check if this is still in the lan prefix? like if i give it `192.168.0.0/30` or something then `101` won't be part of the addr block

File: tstest/natlab/vnet/conf.go (hunk 3 above)
Comment (bradfitz) [reply tstest/natlab/vnet/conf.go:44]: there's enough half done stuff for multiple networks that I'm fine with another here.

File: tstest/natlab/vnet/conf.go (hunk 4 above)
Comment (bradfitz) [reply tstest/natlab/vnet/conf.go:200]: I could require a /24 at least.

Or a /32 but only for the one2one nat mode.

File: tstest/natlab/vnet/nat.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+type NATTable interface {
+	// PickOutgoingSrc returns the source address to use for an outgoing packet.
+	//
+	// The result should either be invalid (to drop the packet) or a WAN (not
+	// private) IP address.
+	//
+	// Typically, the src is a LAN source IP address, but it might also be a WAN
+	// IP address if the packet is being forwarded for a source machine that has
+	// a public IP address.
+	PickOutgoingSrc(src, dst netip.AddrPort, at time.Time) (wanSrc netip.AddrPort)
```
Comment (maisem) [inline tstest/natlab/vnet/nat.go:79]: why `at`? document use case?

Comment (bradfitz) [reply tstest/natlab/vnet/nat.go:79]: I did. Line 67

Comment (maisem) [reply tstest/natlab/vnet/nat.go:79]: ah its on the interface and not on the functions. Maybe repeat?

File: tstest/natlab/vnet/vnet.go [hunk 6]
```diff
@@ -0,0 +1,1120 @@
...
+	"tailscale.com/net/stun"
+	"tailscale.com/syncs"
+	"tailscale.com/tailcfg"
+	"tailscale.com/util/set"
+)
+
+const nicID = 1
+const stunPort = 3478
+
+func (s *Server) PopulateDERPMapIPs() error {
```
Comment (maisem) [inline tstest/natlab/vnet/vnet.go:55]: docs

File: tstest/natlab/vnet/vnet.go [hunk 7]
```diff
@@ -0,0 +1,1120 @@
...
+		for _, n := range r.Nodes {
+			if n.IPv4 != "" {
+				s.derpIPs.Add(netip.MustParseAddr(n.IPv4))
+			}
+		}
+	}
+	return nil
+}
+
+func (n *network) InitNAT(natType NAT) error {
```
Comment (maisem) [inline tstest/natlab/vnet/vnet.go:74]: docs

File: tstest/natlab/vnet/vnet.go [hunk 8]
```diff
@@ -0,0 +1,1120 @@
...
+	const maxInFlightConnectionAttempts = 8192
+	tcpFwd := tcp.NewForwarder(n.ns, tcpReceiveBufferSize, maxInFlightConnectionAttempts, n.acceptTCP)
+	n.ns.SetTransportProtocolHandler(tcp.ProtocolNumber, func(tei stack.TransportEndpointID, pb *stack.PacketBuffer) (handled bool) {
+		log.Printf("TCP packet: %+v", tei)
+		return tcpFwd.HandlePacket(tei, pb)
+	})
+
+	go func() {
+		for {
+			pkt := n.linkEP.ReadContext(n.s.shutdownCtx)
```
Comment (maisem) [inline tstest/natlab/vnet/vnet.go:161]: maybe register something you can wait on for a clean shutdown?

File: tstest/natlab/vnet/vnet.go [hunk 9]
```diff
@@ -0,0 +1,1120 @@
...
+	var wq waiter.Queue
+	ep, err := r.CreateEndpoint(&wq)
+	if err != nil {
+		log.Printf("CreateEndpoint error for %s: %v", stringifyTEI(reqDetails), err)
+		r.Complete(true) // sends a RST
+		return
+	}
+	ep.SocketOptions().SetKeepAlive(true)
+
+	if reqDetails.LocalPort == 123 {
```
Comment (maisem) [inline tstest/natlab/vnet/vnet.go:257]: special port? document

File: tstest/natlab/vnet/vnet.go [hunk 10]
```diff
@@ -0,0 +1,1120 @@
...
+		return
+	}
+
+	var targetDial string
+	if n.s.derpIPs.Contains(destIP) {
+		targetDial = destIP.String() + ":" + strconv.Itoa(int(reqDetails.LocalPort))
+	} else if destIP == fakeControlplaneIP {
+		targetDial = "controlplane.tailscale.com:" + strconv.Itoa(int(reqDetails.LocalPort))
+	}
+	if targetDial != "" {
```
Comment (maisem) [inline tstest/natlab/vnet/vnet.go:271]: other dials?

Comment (maisem) [reply tstest/natlab/vnet/vnet.go:271]: document that they get blackholed or whatever

File: tstest/natlab/vnet/vnet.go [hunk 11]
```diff
@@ -0,0 +1,1120 @@
...
+	mac       MAC
+	portmap   bool
+	wanIP     netip.Addr
+	lanIP     netip.Prefix // with host bits set (e.g. 192.168.2.1/24)
+	nodesByIP map[netip.Addr]*node
+
+	ns     *stack.Stack
+	linkEP *channel.Endpoint
+
+	natStyle syncs.AtomicValue[NAT]
```
Comment (maisem) [inline tstest/natlab/vnet/vnet.go:339]: can this change?

File: tstest/natlab/vnet/vnet.go [hunk 12]
```diff
@@ -0,0 +1,1120 @@
...
+}
+
+type Protocol int
+
+const (
+	ProtocolQEMU      = Protocol(iota + 1)
+	ProtocolUnixDGRAM // for macOS Hypervisor.Framework and VZFileHandleNetworkDeviceAttachment
+)
+
+// serveConn serves a single connection from a client.
```
Comment (maisem) [inline tstest/natlab/vnet/vnet.go:431]: more words, wrong name. Why UnixConn etc

File: tstest/natlab/vnet/vnet.go [hunk 13]
```diff
@@ -0,0 +1,1120 @@
...
+				return
+			}
+			n := binary.BigEndian.Uint32(buf[:4])
+
+			if _, err := io.ReadFull(uc, buf[4:4+n]); err != nil {
+				log.Printf("ReadFull pkt: %v", err)
+				return
+			}
+			packetRaw = buf[4 : 4+n] // raw ethernet frame
+		}
```
Comment (maisem) [inline tstest/natlab/vnet/vnet.go:484]: else?

**Review by maisem — APPROVED** [review]

File: tstest/natlab/vnet/vnet.go (hunk 11 above)
Comment (bradfitz) [reply tstest/natlab/vnet/vnet.go:351]: yes. The plan's to be able to change these at runtime for testing

Comment (maisem) [reply tstest/natlab/vnet/vnet.go:351]: comment?

**Pull request merged**